classes/class-p4-search.php

```diff
@@ -60,28 +83,104 @@ public function __construct( $search_query, $selected_sort = self::DEFAULT_SORT,
 			} else {
 				$this->context = Timber::get_context();
 
+				// Validate user input (sort, filters, etc).
 				if ( $this->validate( $selected_sort, $filters, $this->context ) ) {
 					$this->selected_sort = $selected_sort;
 					$this->filters       = $filters;
 				}
-				$this->posts = $this->get_timber_posts();
+
+				// Set the decoded url query string as key.
+				$query_string = urldecode( filter_input( INPUT_SERVER, 'QUERY_STRING', FILTER_SANITIZE_STRING ) );
+				$group        = 'search';
+				$subgroup     = $this->search_query ? $this->search_query : 'all';
+
+				// Check Object cache for stored key.
+				$this->posts = wp_cache_get( $query_string, "$group:$subgroup" );
```

Comment:
Just a suggestion because I saw this block of code used twice.
You can separate the querying of cache and the adding of key to cache to a separate function.

```php
					$this->posts = wp_cache_get( $query_string, "$group:$subgroup" );
					// TODO - Set the correct filters so that it will work when searching for specific term with filters applied.
					// If cache key expired then retrieve results once again and re-cache them.
					if ( false === $this->posts ) {
						$this->posts = $this->get_timber_posts();
						if ( $this->posts ) {
							wp_cache_add( $query_string, $this->posts, "$group:$subgroup", self::DEFAULT_CACHE_TTL );
						}
					}
```